File: ios/Plugin/LiveKitPlugin.swift
import Foundation
import Capacitor
import Combine
import LiveKitClient

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(LiveKitPlugin)
public class LiveKitPlugin: CAPPlugin {
    public enum Event {
        case callDidStart
        case callDidEnd
        case hang
        case error(Swift.Error)
    }
    private let room = Room()

    private var cancellables = Set<AnyCancellable>()
    
    /// A Combine publisher that clients can subscribe to for API events.
    @objc func connect(_ call: CAPPluginCall) {
        let url = call.getString("url") ?? ""
        let token = call.getString("token") ?? ""
        Task {
            do {
                try await room.connect(url: url, token: token)
                try await room.localParticipant.setCamera(enabled: false);
                try await room.localParticipant.setScreenShare(enabled: false)
                try await room.localParticipant.setMicrophone(enabled: true)
            } catch {
                call.reject(error.localizedDescription )
            }
            call.resolve()
        }
    }
    
    @objc func disconnect(_ call: CAPPluginCall) {
        Task {
            await room.disconnect()
        }
        call.resolve()
    }
    
    @objc func setMuted(_ call: CAPPluginCall) {
        let muted = call.getBool("muted") ?? false
        Task {
            do {
                try await room.localParticipant.setMicrophone(enabled: muted);
                for participant in room.remoteParticipants.values {
                    for publication in participant.audioTracks {
                        if let track = publication.track as? RemoteAudioTrack {
                            track.volume = muted ? 0 : 100
                        }
                    }
                }
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }
}

File: android/src/main/java/com/livekit/capacitor/plugin/LiveKitPlugin.kt
package com.livekit.capacitor.plugin

import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import io.livekit.android.LiveKit.create
import io.livekit.android.LiveKitOverrides
import io.livekit.android.RoomOptions
import io.livekit.android.room.Room
import io.livekit.android.room.track.RemoteTrackPublication
import kotlinx.coroutines.runBlocking

@CapacitorPlugin(name = "LiveKit")
class LiveKitPlugin : Plugin() {
    private var room: Room? = null;

    @PluginMethod
    fun connect(call: PluginCall) {
        if (room == null) {
            room = create(activity.applicationContext)
        }
        val url = call.getString("url")
        val token = call.getString("token")

        runBlocking {
            room!!.connect(url!!, token!!)
            room!!.localParticipant.setCameraEnabled(false)
            room!!.localParticipant.setMicrophoneEnabled(true)
            room!!.localParticipant.setScreenShareEnabled(false)
        }
        call.resolve()
    }

    @PluginMethod
    fun disconnect(call: PluginCall) {
        if (room != null) {
            room!!.disconnect()
        }
        call.resolve()
    }

    @PluginMethod
    fun setMuted(call: PluginCall) {
        val muted = call.getBoolean("muted") ?: false;
        if (room != null) {
            runBlocking {
                room!!.localParticipant.setMicrophoneEnabled(muted)
                for (participant in room!!.remoteParticipants.values) {
                    for (publication in participant.audioTrackPublications) {
                        (publication.first as RemoteTrackPublication).setSubscribed(!muted)
                    }
                }
            }
        }
        call.resolve()
    }
}
